### DISCUSSION TITLE
How to determine if is new user?

### DISCUSSION THREAD

**Discussion by jonwizrd** [nextauthjs/next-auth#1578] [Help]
How to determine new user without setting `session: { jwt: true }`?

**Comment by StefanSelfTaught** [comment]
You can check for the `user.id` in the signIn callback:
```js
async signIn(user, account, profile) {
   if (!user.id) {
      // do something if the user is new
   }
}

**Reply by znwhite** [reply]
@StefanSelfTaught Brand new users still seem to return a temporary `user.id` property.

I tried this:

```
  callbacks: {
    signIn: async (user, account, profile) => {
      if (user.id) {
        console.log('existing user signed in');
      } else {
        console.log('brand new user signed in');
      }
    },
  },
```


Even with a brand new user it still logged "existing user signed in" to the console (using Google login).

**Comment by w0wka91** [comment]
Any news on this?

**Comment by StefanSelfTaught** [comment]
You can now use the message object in the signIn event like this:

```js
events: {
  async signIn(message) {
    console.log(message.isNewUser)
  }
}

**Reply by bennettdams** [reply]
I want to disable sign in for new users, but keep sign in for existing users. As far as I'm aware, there's no "connection" between the `signIn` event and the `signIn` callback.
Is it possible to bring the `isNewUser` property also to the `signIn` callback?

**Reply by GermanJablo** [reply]
I've opened an issue with that request: https://github.com/nextauthjs/next-auth/issues/8272

**Comment by tim-kozak** [comment]
Possible option is to look for registered user fields that are not in provider fields
```
callbacks: {
    signIn: async (user, account, profile) => {
      const isRegistered = user.hasOwnProperty('emailVerified');

      if (isRegistered) {
        console.log('existing user signed in');
      } else {
        console.log('brand new user signed in');
      }
    },
  },
```

**Comment by thander** [comment]
wtf dudes, When will new user detection work? There are different options everywhere and none of them work.
next: 14.2.15
next-auth: 4.24.8

**Reply by thander** [reply]
i found solution
```
  callbacks: {
    async signIn({ user, account, profile, email, credentials }) {
      await connectMongo2()
      const isNewUser = !(await User.findOne({email: user.email}))



      if (isNewUser && !email?.verificationRequest) {

      }

      return true
    },
```
